src/ledger/grain.js

```diff
@@ -185,4 +194,87 @@ export function toFloatRatio(numerator: Grain, denominator: Grain): number {
   return Number(numerator) / Number(denominator);
 }
 
+/**
+ * Splits a budget of Grain into fractional shares.
+ *
+ * It guarantees that the sum of all the splits will be exactly equal
+ * to the budget. It does not guarantee fairness, i.e. some users may
+ * receive more or less Grain than others with equal shares.
+ */
+export function splitBudget(
+  budget: Grain,
+  shares: $ReadOnlyArray<number>
+): $ReadOnlyArray<Grain> {
+  if (lt(budget, ZERO)) {
+    throw new Error("negative budget");
+  }
+  let remainingBudget = budget;
+  let remainingShares = shares.reduce((a, b) => a + b, 0);
+  if (remainingShares === 0) {
+    throw new Error("shares sum to 0");
+  }
+  if (!isFinite(remainingShares)) {
+    throw new Error(`shares sum not finite: ${remainingShares}`);
+  }
+  const pieces = shares.map((s) => {
+    if (s < 0) {
+      throw new Error("negative share: " + s);
+    }
+    if (s === 0 || remainingShares === 0) {
```

Comment:
Isn't the `remainingShares === 0` case already handled on line 213?

Reply:
No, in line 213 we haven't started mutating remainingShares, so at that point it's effectively the total shares.

Reply:
got it, didnt see the mutation below.

Reply:
Tweaked the implementation to make this clearer.